_extensions/custom_roles.py

```diff
@@ -19,12 +25,16 @@ def setup(app):
 
 def autolink(pattern):
     def role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+        """Combine literal + reference (unless the text is prefixed by a !)."""
         if " " in text:
             url_text = urllib.parse.quote(f"{text}")
         else:
             url_text = text
-        url = pattern % (url_text,)
-        node = nodes.reference(rawtext, text, refuri=url, **options)
+        url = pattern.format(url_text)
+        node = nodes.literal(rawtext, text.lstrip('!'), **options)
+        # don't create a reference if the text starts with !
+        if not text.startswith('!'):
+            node = nodes.reference('', '', node, refuri=url, **options)
```

Comment:
The current implementation disallows e.g. ``:cpy-file:`!!spam!.txt` `` (I guess this is never likely to actually be needed). Refactored to skip options per above, and pass internal=False/rawtext to the reference node.
```suggestion
        # don't create a reference if the text starts with !
        if text.startswith('!'):
            node = nodes.literal(rawtext, text[1:])
        else:
            node = nodes.reference(rawtext, '', nodes.literal(rawtext, text),
			                       refuri=url, internal=False)
```